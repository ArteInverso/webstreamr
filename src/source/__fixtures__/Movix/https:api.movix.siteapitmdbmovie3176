{"tmdb_details":{"id":3176,"title":"Battle Royale","original_title":"バトル・ロワイアル","release_date":"2000-12-16","poster_path":"/laUmjb39U6glMrOKyaco4QPeGaC.jpg","backdrop_path":"/amBvmIshdsSkOtvVIgxl7YSQ9Dg.jpg","overview":"Dans un avenir proche, les élèves de la classe B de 3ème du collège Shiroiwa ont été amenés sur une île déserte par une armée mystérieuse. Un adulte surgit tout à coup devant eux : leur ancien professeur Kitano. Il leur annonce qu'ils vont participer à un jeu de massacre dont la règle consiste à s'entretuer. Seul le dernier des survivants pourra regagner son foyer. Kitano leur présente deux nouveaux élèves très inquiétants. Des coups de feu retentissent pour convaincre les incrédules. Selon la loi de réforme de l'éducation pour le nouveau siècle, ce sacrifice permettra de former des adultes sains. Abandonnés chacun à son sort avec de la nourriture et une arme, les adolescents disposent d'un délai de trois jours pour s'entretuer.","vote_average":7.284},"iframe_src":"https://lecteurvideo.com/embed.php?id=17959&ads=true","player_links":[{"decoded_url":"https://lecteur6.com/video/18085327b86002fc604c323b9a07f997","quality":"Vidéo HD / PAS DE PUBLICITE 2","language":"French"},{"decoded_url":"https://vidoza.net/embed-55hvynbdyqvc.html","quality":"Vid Vidéo 6 / PAS DE PUBLICITE","language":"French"},{"decoded_url":"https://wishonly.site/e/kykhuv6tscs8","quality":"Wish Vidéo 4 / PAS DE PUBLICITE","language":"French"},{"decoded_url":"https://darkibox.com/embed-5ghz7myucfjq.html","quality":"Darki Vidéo HD / PAS DE PUBLICITE","language":"French"},{"decoded_url":"https://supervideo.cc/e/uieh17bid0ce","quality":"Super Vidéo HD / PAS DE PUBLICITE","language":"French"},{"decoded_url":"https://dood.li/e/gha75crrm8nf","quality":"Doodstream Vidéo 9 / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://filemoon.sx/e/c5lhlypfasmm","quality":"FileMon Vidéo 10 / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://voe.sx/e/aghpmosgkhsu","quality":"VoeSx Vidéo 11 / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"http://vidmoly.me/embed-79w00r2qv88n.html","quality":"VidMoly Vidéo 12 / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://waaw.to/f/JWzt5eJSbpu8","quality":"Netu Vidéo 14 / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://veev.to/e/lu4c24e3aetf","quality":"Veed / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://listeamed.net/e/eL2157Kzw8m5rw4","quality":"VidGuard / Le serveur vidéo contient des publicités","language":"French"}]}