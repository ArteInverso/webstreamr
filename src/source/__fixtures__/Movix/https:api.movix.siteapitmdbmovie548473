{"message":"Contenu non disponible","tmdb_id":"548473","tmdb_details":{"id":548473,"title":"La couleur tombée du ciel","original_title":"Color Out of Space","release_date":"2020-01-24","poster_path":"/nVwMu8nExJIsnQt1hheoBd85wdR.jpg","backdrop_path":"/iyn2JQsftaqhQlXYaCCvn8udM5i.jpg","overview":"Après la chute d’une météorite dans le jardin de leur ferme, Nathan Gardner et sa famille se retrouvent à combattre un organisme extraterrestre et mutant qui s’est introduit dans leurs corps et leurs cerveaux. La quiétude de leur vie à la campagne va se transformer en cauchemar en technicolor.","vote_average":6}}