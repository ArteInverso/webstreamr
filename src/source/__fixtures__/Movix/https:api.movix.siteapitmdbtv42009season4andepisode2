{"tmdb_details":{"id":42009,"title":"Black Mirror","original_title":"Black Mirror","release_date":"2011-12-04","poster_path":"/9acfIYfBuB4GFVROipM9YrqxsXd.jpg","backdrop_path":"/dg3OindVAGZBjlT3xYKqIAdukPL.jpg","overview":"Les histoires tordues se succèdent dans cette série d'anthologie hallucinante qui révèle les plus grandes innovations de l'humanité, mais aussi ses côtés les plus sombres.","vote_average":8.284},"seasons":[{"season_number":1,"name":"Black Mirror - Season 1","data_id":"73065","post_id":"19799","episodes":[]},{"season_number":2,"name":"Black Mirror - Season 2","data_id":"73066","post_id":"19799","episodes":[]},{"season_number":3,"name":"Black Mirror - Season 3","data_id":"73067","post_id":"19799","episodes":[]},{"season_number":4,"name":"Black Mirror - Season 4","data_id":"73068","post_id":"19799","episodes":[]},{"season_number":5,"name":"Black Mirror - Season 5","data_id":"73069","post_id":"19799","episodes":[]},{"season_number":6,"name":"Black Mirror - Season 6","data_id":"73070","post_id":"19799","episodes":[]},{"season_number":7,"name":"Black Mirror - Season 7","data_id":"229573","post_id":"19799","episodes":[]}],"current_episode":{"season_number":4,"episode_number":2,"title":"Black Mirror","iframe_src":"https://lecteurvideo.com/embed.php?id=17320&ads=true","player_links":[{"decoded_url":"https://uqload.net/embed-xjmhb7rbzdjh.html","quality":"Uqload Vidéo 8 / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://lulustream.com/e/c3defa2cbtmm","quality":"LuluStream / PAS DE PUBLICITE","language":"French"},{"decoded_url":"https://wishonly.site/e/bzfdmjlrsqxk","quality":"Wish Vidéo 4 / PAS DE PUBLICITE","language":"French"},{"decoded_url":"https://vidoza.net/embed-i2e8gn4m64e1.html","quality":"Vid Vidéo 6 / PAS DE PUBLICITE","language":"French"},{"decoded_url":"https://filemoon.sx/e/0iuzfpt5cbx6","quality":"FileMon Vidéo 10 / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://voe.sx/e/mustkdlxo67t","quality":"VoeSx Vidéo 11 / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"http://vidmoly.me/embed-2rls953tvhxc.html","quality":"VidMoly Vidéo 12 / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://waaw.to/f/GdQ6mHpxPdd8","quality":"Netu Vidéo 14 / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://veev.to/e/nifwerilwo0i","quality":"Veed / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://listeamed.net/e/8ozgENePwdLxmjA","quality":"VidGuard / Le serveur vidéo contient des publicités","language":"French"},{"decoded_url":"https://coflix.upn.one/#x63hig","quality":"coflix.upn","language":"French"},{"decoded_url":"https://do7go.com/e/2zn4y7ueq1ta","quality":"do7go","language":"French"},{"decoded_url":"https://movearnpre.com/embed/eiel2p4z74kp","quality":"movearnpre","language":"French"}]}}