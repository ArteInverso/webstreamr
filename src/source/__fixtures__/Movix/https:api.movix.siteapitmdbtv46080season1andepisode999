{"message":"Contenu non disponible","tmdb_id":"46080","tmdb_details":{"id":46080,"title":"Masha et Michka","original_title":"Маша и Медведь","release_date":"2009-01-07","poster_path":"/ruQSBtzk5d6mQHLoiBvOQEr728R.jpg","backdrop_path":"/uu2qZZ82Ml2u6SgOd68OIIOI0si.jpg","overview":"Masha est une petite fille qui n'arrive pas facilement à se poser tant son énergie est grande. Sa curiosité et sa créativité provoquent des situations comiques, même si parfois cela crée des problèmes aux autres. L'ours, Michka, est un ami au coeur tendre qui voit son monde tranquille bouleversé dès que Masha entre en action...","vote_average":6.9}}