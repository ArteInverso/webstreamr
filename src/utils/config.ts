import { Config } from '../types';

export const getDefaultConfig = (): Config => {
  return { multi: 'on', en: 'on' };
};

export const showExternalUrls = (config: Config): boolean => 'includeExternalUrls' in config;

export const hasMultiEnabled = (config: Config): boolean => 'multi' in config;
