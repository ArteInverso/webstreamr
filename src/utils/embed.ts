import { unpack } from 'unpacker';
import { cachedFetchText } from './fetch';
import { scanFromResolution } from './resolution';

const LINK_REGEXPS = [
  /sources:\[{file:"(.*?)"/, // dropload, supervideo
];

interface ParsePackedEmbedResult {
  url: string;
  resolution: string | undefined;
  size: string | undefined;
}

const findResolution = (html: string): string | undefined => {
  const match = html.match(/(\d{3,}x\d{3,}),/);

  return match && match[1] ? scanFromResolution(match[1]) : undefined;
};

const findSize = (html: string): string | undefined => {
  const sizeMatch = html.match(/([\d.]+) ?([GM]B)/);

  return sizeMatch && sizeMatch[1] && sizeMatch[2] ? `${sizeMatch[1]} ${sizeMatch[2]}` : undefined;
};

export const parsePackedEmbed = async (url: string): Promise<ParsePackedEmbedResult> => {
  const html = await cachedFetchText(url);

  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found on ${url}`);
  }

  const unpacked = unpack(evalMatch[0]);

  for (const linkRegexp of LINK_REGEXPS) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return {
        url: 'https://' + linkMatch[1].replace(/^(https:)?\/\//, ''),
        resolution: findResolution(html),
        size: findSize(html),
      };
    }
  }

  throw new Error(`Could not find a stream link on ${url}`);
};
