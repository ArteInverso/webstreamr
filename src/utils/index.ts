export * from './env';
export * from './Fetcher';
export * from './StreamResolver';
export * from './embed';
export * from './imdb';
export * from './languageFromCountryCode';
export * from './manifest';
export * from './tmdb';
