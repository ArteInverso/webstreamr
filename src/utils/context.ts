import { Request } from 'express';
import { Context } from '../types';
import { getDefaultConfig } from './config';

export const contextFromRequest = (req: Request): Context => {
  return {
    hostUrl: new URL(`${req.protocol}://${req.host}`),
    id: req.headers['X-Request-ID'] as string,
    ...(req.ip && { ip: req.ip }),
    config: req.params['config'] ? JSON.parse(req.params['config']) : getDefaultConfig(),
  };
};
