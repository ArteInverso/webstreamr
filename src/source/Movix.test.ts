import { createTestContext } from '../test';
import { FetcherMock, TmdbId } from '../utils';
import { Movix } from './Movix';

const ctx = createTestContext({ fr: 'on' });

describe('Movix', () => {
  let handler: Movix;

  beforeEach(() => {
    handler = new Movix(new FetcherMock(`${__dirname}/__fixtures__/Movix`));
  });

  test('handles non-existent content gracefully', async () => {
    const streams = await handler.handle(ctx, 'series', new TmdbId(46080, 1, 999));
    expect(streams).toMatchSnapshot();
  });

  test('handle tmdb black mirror s4e2', async () => {
    const streams = await handler.handle(ctx, 'series', new TmdbId(42009, 4, 2));
    expect(streams).toMatchSnapshot();
  });

  test('handle battle royal', async () => {
    const streams = await handler.handle(ctx, 'movie', new TmdbId(3176, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });
});
