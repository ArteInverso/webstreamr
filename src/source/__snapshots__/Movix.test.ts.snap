// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Movix handle battle royal 1`] = `
[
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://lecteur6.com/video/18085327b86002fc604c323b9a07f997",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://vidoza.net/embed-55hvynbdyqvc.html",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://wishonly.site/e/kykhuv6tscs8",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://darkibox.com/embed-5ghz7myucfjq.html",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://supervideo.cc/e/uieh17bid0ce",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://dood.li/e/gha75crrm8nf",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://filemoon.sx/e/c5lhlypfasmm",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://voe.sx/e/aghpmosgkhsu",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "http://vidmoly.me/embed-79w00r2qv88n.html",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://waaw.to/f/JWzt5eJSbpu8",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://veev.to/e/lu4c24e3aetf",
  },
  {
    "countryCode": "fr",
    "title": "Battle Royale",
    "url": "https://listeamed.net/e/eL2157Kzw8m5rw4",
  },
]
`;

exports[`Movix handle tmdb black mirror s4e2 1`] = `
[
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://uqload.net/embed-xjmhb7rbzdjh.html",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://lulustream.com/e/c3defa2cbtmm",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://wishonly.site/e/bzfdmjlrsqxk",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://vidoza.net/embed-i2e8gn4m64e1.html",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://filemoon.sx/e/0iuzfpt5cbx6",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://voe.sx/e/mustkdlxo67t",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "http://vidmoly.me/embed-2rls953tvhxc.html",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://waaw.to/f/GdQ6mHpxPdd8",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://veev.to/e/nifwerilwo0i",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://listeamed.net/e/8ozgENePwdLxmjA",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://coflix.upn.one/#x63hig",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://do7go.com/e/2zn4y7ueq1ta",
  },
  {
    "countryCode": "fr",
    "title": "Black Mirror 4x2",
    "url": "https://movearnpre.com/embed/eiel2p4z74kp",
  },
]
`;

exports[`Movix handles non-existent movie gracefully 1`] = `[]`;

exports[`Movix handles non-existent show gracefully 1`] = `[]`;
